File: 2498-smallest-subarrays-with-maximum-bitwise-or/2498-smallest-subarrays-with-maximum-bitwise-or.java
class Solution {
    public int [] smallestSubarrays (int [] nums) {
        int last [] = new int [32];
        for (int i = nums.length - 1; i >= 0; i--) {
            int maxIndex = i;
            for (int b = 0; b < 32; b++) {
                if ((nums [i] & (1 << b)) != 0)
                    last [b] = i;
                else
                    maxIndex = Math.max (maxIndex, last [b]);
            }
            nums [i] = maxIndex - i + 1;
        }
        return nums;
    }
}

File: 2498-smallest-subarrays-with-maximum-bitwise-or/2498-smallest-subarrays-with-maximum-bitwise-or.c
/**
 * Note: The returned array must be malloced, assume caller calls free().
 */
int* smallestSubarrays (int* nums, int numsSize, int* returnSize) {
    int last [32] = {0};
    for (int i = numsSize - 1; i >= 0; i--){
        int maxIndex = i;
        for (int b = 0; b < 32; b++){
            if (nums [i] & (1L << b))
                last [b] = i;
            else
                maxIndex = fmax (maxIndex, last[b]);
        }
        nums [i] = maxIndex - i + 1;
    }
    *returnSize = numsSize;
    return nums;
}